FIX: application update command should update packaged app manifest and content

rpl opened the pull request:
- this change should fix #269
- introduce a temporary webappsActor renamed to simulatorwebappsActor to fix a couple of issues with reinstall application using its install command
- kill the app if it's already running to run the new version correctly
- install app if not installed on run button press 
- reinstall apps on first run after upgrade/downgrade/remote+install (fix #308)

rpl committed:
simulatorActor runApp command: kill app by origin before the run app step

rpl committed:
FIX: add a temporary fixed dbg-webapps-actor.js as simulatorwebappsActor

this patch could be reverted once this fix (or another solution to the same
problem) will be introduced into the dbg-webapps-actor from the b2g repo.

rpl committed:
simulatorActor runApp command: workaround unlock screen problem

(often gaia is not fully loaded and it isn't ready to handle the unlock screen request)

rpl commented on addon/lib/remote-simulator-client.js:
```diff
@@ -76,7 +76,7 @@ const RemoteSimulatorClient = Class({
           selected: reply.selected,
           simulator: reply.simulatorActor,
-          webapps: reply.webappsActor
+          webapps: reply.simulatorwebappsActor
```
use the temporary fixed webappsActor renamed to simulatorwebappsActor

mykmelez replied:
Nit: for names like this, capitalize the previously-initial character of the old name (i.e. simulatorWebappsActor) or separate the new prefix from the old name with an underscore character (i.e. simulator_webappsActor).

rpl commented on prosthesis/content/dbg-webapps-actors.js:
```diff
@@ -0,0 +1,281 @@
+    let reg = DOMApplicationRegistry;
+    let self = this;
+
+    if (aId in reg._manifestCache) {
```
clean the cached manifest (e.g. delete DOMApplicationRegistry._manifestCache[appId]) as done here: http://mxr.mozilla.org/mozilla-central/source/dom/apps/src/Webapps.jsm#1084

rpl commented on prosthesis/content/dbg-webapps-actors.js:
```diff
@@ -0,0 +1,281 @@
+          zipFile.append("application.zip");
+
+          // Refresh applicazion.zip content (e.g. reinstall app)
+          Services.obs.notifyObservers(zipFile, "flush-cache-entry", null);
```
flush zip reader cache  as done here: http://mxr.mozilla.org/mozilla-central/source/dom/apps/src/Webapps.jsm#1117

rpl commented on prosthesis/content/simulator-actors.js:
```diff
@@ -63,43 +63,62 @@ SimulatorActor.prototype = {
-        let setReq = window.navigator.mozSettings
-          .createLock().set({'lockscreen.enabled': false});
-        setReq.onsuccess = function() {
+        // WORKAROUND: currently we're not able to detect when the firefoxos is fully loaded
```
I would like a better workaround to this problem :-(

mykmelez replied:
Can we disable the lockscreen permanently, f.e. via build/override-settings.json? It isn't useful for app developers right now, it's just a hoop to jump through. And in the rare case where a developer really needs to see the lockscreen, presumably they can do so by pressing the End key (Fn+RightArrow on Mac).

rpl commented on prosthesis/content/simulator-actors.js:
```diff
@@ -63,43 +63,62 @@ SimulatorActor.prototype = {
+      killAppByOrigin: function(origin, cb) {
+        try {
+          WindowManager.kill(origin);
+          // WORKAROUND: currently WindowManager.kill doesn't always call
```
I would like a better workaround to this problem :-( :-( :-(

mykmelez replied:
Check `(WindowManager.getRunningApps().hasOwnProperty[origin] || !WindowManager.getRunningApps()[origin].killed)` before killing the app, then listen for an _appterminated_ event instead of passing a callback.

_kill()_ only returns early if `!isRunning(origin)` or `runningApps[origin].killed`. And _isRunning()_ just checks `runningApps.hasOwnProperty(origin)`. Where _runningApps_ is the object returned by _getRunningApps()_. It still doesn't always call the callback, even if it doesn't return early. But it does always dispatch the event if it doesn't return early.

mykmelez commented on prosthesis/content/dbg-webapps-actors.js:
```diff
@@ -0,0 +1,281 @@
+          zipFile = installDir.clone();
+          zipFile.append("application.zip");
+
+          // Refresh applicazion.zip content (e.g. reinstall app)
```
Nit: applicazion -> application

rpl committed:
rename temporary simulatorwebappsActor to simulatorWebappsActor

rpl committed:
disable lockscreen on simulator b2g-desktop instances

rpl committed:
fixed typo

rpl committed:
unlock screen on b2g-desktop start and set homescreen.ready settings when homescreen is fully loaded

rpl committed:
waitHomescreenReady helper on runApp

rpl committed:
tweaks on lockscreen disable, homescreen ready notify and simulator-actors runApp command

rpl committed:
remove trailing spaces

rpl committed:
rename simulator-actors to dbg-simulator.actors and extract lockscreen.js from shell.js

rpl committed:
fix trailing spaces

rpl commented:
renamed prosthesis/content/simulator-actors.js to prosthesis/content/dbg-simulator-actors.js
(follow debugging actors definition files naming conventions)

rpl commented on prosthesis/content/dbg-simulator-actors.js:
```diff
@@ -63,43 +63,93 @@ SimulatorActor.prototype = {
+          log("RUNAPP EXCEPTION: " + e);
+        }
+      },
+      waitHomescreenReady: function(cb) {
```
wait for "homescreen.ready" setting to be true:
- b2g instance already started
- b2g instance starting, lockscreen is fully loaded and correctly unlocked

rpl commented on prosthesis/content/dbg-simulator-actors.js:
```diff
@@ -63,43 +63,93 @@ SimulatorActor.prototype = {
+              homescreen.removeEventListener("appterminated", once);
+              // WORKAROUND: bug (probably related to disabled oop) restarted
+              // app keep to be flagged as killed and never restarted
+              delete WindowManager.getRunningApps()[origin];
```
I'm not sure why happening, but after a couple of times a user click the "update" app button,
window_manager keeps to terminated app as killed in the running apps list and never restarts
it again:
- !!WindowManager.getRunningApps()[origin] == true
- WindowManager.getRunningApps()[origin].killed == true

rpl commented on prosthesis/content/lockscreen.js:
```diff
@@ -0,0 +1,39 @@
+
+// NOTE: disable lockscreen by default on FirefoxOS Simulator
+// disable lockscreen on startup
+SettingsListener.observe("lockscreen.enabled", false, function(value) {
```
lockscreen.enabled is set and never read by LockScreen,
we use it to detect when lockscreen start loading.

rpl commented on prosthesis/content/lockscreen.js:
```diff
@@ -0,0 +1,39 @@
+  let homescreen = document.getElementById("homescreen").contentWindow.wrappedJSObject;
+  dump("simulator - HOMESCREEN LOCKED: wait for lockscreen.locked \n");
+  // keep the lockscreen unlocked
+  SettingsListener.observe("lockscreen.locked", true, function(value) {
```
lockscreen.locked is set by LockScreen on lock/unlock

this handler keep the lockscreen unlocked and set "homescreen.ready = true"

rpl committed:
remove old workaround (and disabled) code

rpl committed:
runApp and uninstall should use appId + install app if not installed on runApp user request

change remote-simulator-client runApp and uninstall methods to use appId
(an in the install method) and use the appId to check if the application
is already installed in the DOMApllicationRegistry.webapps on the b2g side.

If an app in not installed try to install and run it again.

rpl committed:
on upgrade ensureXkeysUnique and leave apps to be lazily reinstalled by the run button (should fix #308)

rpl commented on addon/lib/simulator.js:
```diff
@@ -764,13 +764,26 @@ let simulator = module.exports = {
+        simulator.runApp(app, function (res) {
+          if (res.success === false) {
+            if (res.error === 'app-not-installed') {
+              // install and run if not installed
```
we're now able to detect if an app is not installed, so we can try to install and run it again.
thanks to this change we can lazily reinstalled an app when:
- user remove the app manually from the homescreen
- user upgrade or reinstall the simulator addon

reinstalling the addon xpi will overwrite the b2g profile and all the app registered in the dashboard
will became "not installed".

rpl commented on addon/lib/main.js:
```diff
@@ -76,8 +78,8 @@ switch (Self.loadReason) {
-
-    Simulator.updateAll();
+    // NOTE: updateAll disabled (lazy install if not installed on runApp)
+    //Simulator.updateAll();
```
thanks to the "install if not installed" feature in the simulator runApp method,
we can now alleviate #308 ensuring xkey uniqueness on upgrade and leave
apps to lazily reinstall when the user press the run button.

rpl committed:
FIX: evict b2g Webapps.js component's private manifest cache on application install

this change is needed to propagate manifest changes to the homescreen gaia application
on app reinstall without a b2g instance restart.

rpl committed:
disable screen timeout

we don't need this power saving settings to be enabled on the b2g instance
running on the simulator.

mykmelez commented:
Oh, and thanks for removing the trailing spaces and unused code! In the long run, it'll make a difference to keep our code tidy in these and other ways.

rpl committed:
fix coding style and add more comments

rpl committed:
reinstall active registered apps on first b2g-desktop run (fix #308)

rpl committed:
showNotification on b2g instance to inform user of the reinstall process

rpl commented on addon/lib/main.js:
```diff
@@ -62,23 +62,26 @@ if (SStorage.storage.lastVersion != Self.version) {
+// on remove+install, downgrade and upgrade:
+// - ensure apps xkeys are unique
+// - flag needsUpdateAll if there are active apps registered
+if (["install", "downgrade", "upgrade"].indexOf(Self.loadReason) >= 0) {
```
on remove+install, downgrade and upgrade we just ensure apps xkeys are unique and flag needsUpdateAll if there are active apps registered
("deleted" apps should not be reinstalled)

rpl commented on addon/lib/simulator.js:
```diff
@@ -131,13 +131,14 @@ let simulator = module.exports = {
 
-  updateAll: function() {
+  updateAll: function(oncompleted) {
+    simulator.remoteShowNotification("reinstalling registered apps...");
```
add notification on b2g side to keep the user informed of reinstall apps procedure

rpl commented on addon/lib/simulator.js:
```diff
@@ -148,14 +149,29 @@ let simulator = module.exports = {
             iterator.next();
-          } catch (err if err instanceof StopIteration) {}
+          } catch (err if err instanceof StopIteration) {
+            simulator.remoteShowNotification("reinstalling completed.");
```
add another notification on b2g side to signal completed reinstall apps procedure

rpl commented on addon/lib/simulator.js:
```diff
@@ -148,14 +149,29 @@ let simulator = module.exports = {
         }, 0);
       }
-      let iterator = (simulator.updateApp(id, next) for (id in simulator.apps));
+      // only active apps needs to be reinstalled
```
only active apps needs to be reinstalled

rpl commented on addon/lib/simulator.js:
```diff
@@ -677,14 +693,27 @@ let simulator = module.exports = {
 
+    let next = null;
+    // if needsUpdateAll try to reinstall all active registered app
+    if (SStorage.storage.needsUpdateAll) {
```
if needsUpdateAll try to reinstall all active registered app

rpl commented on prosthesis/content/dbg-simulator-actors.js:
```diff
@@ -154,7 +220,19 @@ SimulatorActor.prototype = {
+    };
+  },
+
+  onShowNotification: function (aRequest) {
```
use AlertHelper to show notification on the b2g side

rpl committed:
rename simulator.remoteShowNotification to simulator.showRemoteNotification

rpl committed:
onResponse is optional, pass null to showNotification instead of an empty callback

rpl committed:
fix coding syle

rpl committed:
FIX: if needsUpdatesAll is true updateAll should be called on ready

rpl committed:
fix typos

mykmelez commented on addon/lib/simulator.js:
```diff
@@ -148,14 +149,29 @@ let simulator = module.exports = {
+  showRemoteNotification: function(userMessage) {
+    this.run(function () {
+      simulator.remoteSimulator
+        .showNotification(userMessage, null);
```
Nit: since the _onResponse_ argument is optional, just leave it out instead of passing _null_.

mykmelez commented:
Merged with nit fixed!
